### ISSUE TITLE
Functions for looking up previous/next/all occurrences

### ISSUE THREAD

**Issue by tysonzero** [mitchellvitez/rrule#2]
These would be extremely useful for working with recurrence rules.

**Comment by tysonzero** [comment]
I'm currently using the `iCalendar` library, but with this additional functionality I'll either use them in parallel or ask the `iCalendar` library to build on top of this one.

**Comment by mitchellvitez** [comment]
@tysonzero I would like to add this `expand` function, would it fit your use case?

```
expand :: RRule -> UTCTime -> [UTCTime]
expand rule startTime = _
```

`next :: RRule -> UTCTime -> Maybe UTCTime` would then just need to grab the first item from an `expand`

I haven't thought much about `previous` yet. Would that be the only useful bit, or would some kind of `expandBackwardsInTime` (hopefully with a better name) be useful in your case?

**Comment by mitchellvitez** [comment]
Thanks to my coworker @lykahb for some help in thinking about this.

I believe the `libical` C library is what one would want to use to build this functionality. See [libical/icalrecur.c](https://github.com/libical/libical/blob/master/src/libical/icalrecur.c) for the relevant logic.

I wouldn't want to do a Haskell re-write of that logic, since that seems bug-prone. However, I also don't want to pull in `libical` as a dependency here, since it seems kind of out of scope for what I intended RRule to do. Maybe the `iCalendar` Haskell library you mentioned would be okay with pulling it in?

---

Despite not wanting to add this change, I did enough research about it that I think I might be able to be helpful.

Let's assume you want to use `libical` directly. If you call `icalrecur_expand_recurrence` over FFI, you get a function like

`icalrecur_expand_recurrence :: CString -> CTime -> CInt -> Ptr CTime -> IO CInt`

From this, it's possible to write a function getting you the possibly-infinite list of recurrences coming after some `UTCTime`

```hs
expand :: RRule -> UTCTime -> [UTCTime]
```

There are a couple obvious ways to add usefulness to this; you'd get `next` for cheap

```hs
next :: RRule -> UTCTime -> Maybe UTCTime
next rule time = listToMaybe $ expand rule time
```

or, for example, getting all recurrences between two `UTCTime`s

```hs
between :: RRule -> UTCTime -> UTCTime -> [UTCTime]
between rule start end = takeWhile (< end) $ expand rule start
```

Unfortunately I'm not sure offhand how to implement `previous`. Perhaps there's something in `libical` that provides this.
